File: DSP/lib/audio/wav_utils.dart
/// Pure-Dart WAV file reader/writer for 16-bit PCM audio.
///
/// No external packages — uses only `dart:io` and `dart:typed_data`.
/// Handles standard RIFF/WAVE/fmt/data chunks. Reads mono or stereo
/// (stereo is downmixed to mono by averaging channels).
library;

import 'dart:io';
import 'dart:typed_data';

/// Container for WAV sample data and its sample rate.
class WavData {
  /// Audio samples normalised to the range [-1.0, 1.0].
  final List<double> samples;

  /// The sample rate read from the WAV file header.
  final int sampleRate;

  WavData(this.samples, this.sampleRate);
}

class WavUtils {
  /// Reads a 16-bit PCM mono or stereo WAV from [bytes] and returns
  /// its samples as normalised doubles in [-1.0, 1.0].
  ///
  /// If stereo, downmixes to mono by averaging left and right channels.
  /// Throws [ArgumentError] if the bytes do not represent a valid 16-bit PCM WAV.
  static WavData readWavBytes(Uint8List bytes) {
    final data = ByteData.sublistView(bytes);

    // --- RIFF header ---
    if (bytes.length < 12) {
      throw ArgumentError('WAV byte buffer too short: ${bytes.length} bytes');
    }
    final riff = String.fromCharCodes(bytes.sublist(0, 4));
    if (riff != 'RIFF') {
      throw ArgumentError('Not a RIFF buffer');
    }
    final wave = String.fromCharCodes(bytes.sublist(8, 12));
    if (wave != 'WAVE') {
      throw ArgumentError('Not a WAVE buffer');
    }

    // --- Walk sub-chunks to find "fmt " and "data" ---
    int? fmtOffset;
    int? dataOffset;
    int? dataSize;
    int offset = 12; // right after "WAVE"
    while (offset + 8 <= bytes.length) {
      final chunkId = String.fromCharCodes(bytes.sublist(offset, offset + 4));
      final chunkSize = data.getUint32(offset + 4, Endian.little);
      if (chunkId == 'fmt ') {
        fmtOffset = offset + 8;
      } else if (chunkId == 'data') {
        dataOffset = offset + 8;
        dataSize = chunkSize;
      }
      // Move to next chunk (chunks are word-aligned)
      offset += 8 + chunkSize;
      if (offset % 2 != 0) offset++;
    }

    if (fmtOffset == null) {
      throw ArgumentError('Missing "fmt " chunk in WAV buffer');
    }
    if (dataOffset == null || dataSize == null) {
      throw ArgumentError('Missing "data" chunk in WAV buffer');
    }

    // --- Parse fmt chunk ---
    final audioFormat = data.getUint16(fmtOffset, Endian.little);
    if (audioFormat != 1) {
      throw ArgumentError(
        'Unsupported audio format $audioFormat (only PCM = 1 is supported)',
      );
    }
    final numChannels = data.getUint16(fmtOffset + 2, Endian.little);
    if (numChannels < 1 || numChannels > 2) {
      throw ArgumentError('Unsupported channel count $numChannels (only 1 or 2 supported)');
    }
    final sampleRate = data.getUint32(fmtOffset + 4, Endian.little);
    if (sampleRate <= 0) {
      throw ArgumentError('Invalid sample rate $sampleRate');
    }
    final bitsPerSample = data.getUint16(fmtOffset + 14, Endian.little);
    if (bitsPerSample != 16) {
      throw ArgumentError(
        'Unsupported bit depth $bitsPerSample (only 16-bit PCM is supported)',
      );
    }
    if (dataOffset + dataSize > bytes.length) {
      throw ArgumentError('Declared data chunk size exceeds available buffer length');
    }

    // --- Read PCM samples ---
    const bytesPerSample = 2; // 16-bit
    final totalSampleValues = dataSize ~/ bytesPerSample;
    final totalFrames = totalSampleValues ~/ numChannels;

    final samples = List<double>.filled(totalFrames, 0.0);
    for (int frame = 0; frame < totalFrames; frame++) {
      double sum = 0.0;
      for (int ch = 0; ch < numChannels; ch++) {
        final sampleOffset =
            dataOffset + (frame * numChannels + ch) * bytesPerSample;
        final int16Value = data.getInt16(sampleOffset, Endian.little);
        sum += int16Value / 32768.0;
      }
      samples[frame] = sum / numChannels; // average channels for mono mix
    }

    return WavData(samples, sampleRate);
  }

  /// Reads a 16-bit PCM mono or stereo WAV file from [path] and returns
  /// its samples as normalised doubles in [-1.0, 1.0], at whatever sample
  /// rate the file actually has.
  static WavData readWav(String path) {
    final file = File(path);
    final bytes = file.readAsBytesSync();
    return readWavBytes(Uint8List.fromList(bytes));
  }

  /// Encodes [samples] (normalised doubles in [-1.0, 1.0]) as a 16-bit PCM
  /// mono WAV byte buffer at [sampleRate].
  ///
  /// Clips values to [-1.0, 1.0] before converting to avoid integer
  /// overflow/wraparound artifacts.
  static Uint8List writeWavBytes(List<double> samples, {required int sampleRate}) {
    const numChannels = 1;
    const bitsPerSample = 16;
    const bytesPerSample = bitsPerSample ~/ 8;
    final dataSize = samples.length * bytesPerSample;
    const fmtChunkSize = 16;
    // Total file size = 4 (RIFF type) + 8+fmtChunkSize (fmt chunk)
    //                  + 8+dataSize (data chunk)
    final fileSize = 4 + (8 + fmtChunkSize) + (8 + dataSize);

    final buffer = ByteData(8 + fileSize); // 8 = "RIFF" + file size field
    int pos = 0;

    // --- RIFF header ---
    void writeString(String s) {
      for (int i = 0; i < s.length; i++) {
        buffer.setUint8(pos++, s.codeUnitAt(i));
      }
    }

    writeString('RIFF');
    buffer.setUint32(pos, fileSize, Endian.little);
    pos += 4;
    writeString('WAVE');

    // --- fmt sub-chunk ---
    writeString('fmt ');
    buffer.setUint32(pos, fmtChunkSize, Endian.little);
    pos += 4;
    buffer.setUint16(pos, 1, Endian.little); // PCM
    pos += 2;
    buffer.setUint16(pos, numChannels, Endian.little);
    pos += 2;
    buffer.setUint32(pos, sampleRate, Endian.little);
    pos += 4;
    buffer.setUint32(
      pos,
      sampleRate * numChannels * bytesPerSample,
      Endian.little,
    ); // byte rate
    pos += 4;
    buffer.setUint16(
      pos,
      numChannels * bytesPerSample,
      Endian.little,
    ); // block align
    pos += 2;
    buffer.setUint16(pos, bitsPerSample, Endian.little);
    pos += 2;

    // --- data sub-chunk ---
    writeString('data');
    buffer.setUint32(pos, dataSize, Endian.little);
    pos += 4;

    for (final sample in samples) {
      // Clip to [-1.0, 1.0] then convert to int16
      final clamped = sample.clamp(-1.0, 1.0);
      final int16Value = (clamped * 32767.0).round().clamp(-32768, 32767);
      buffer.setInt16(pos, int16Value, Endian.little);
      pos += 2;
    }

    return buffer.buffer.asUint8List(0, pos);
  }

  /// Writes [samples] (normalised doubles in [-1.0, 1.0]) as a 16-bit PCM
  /// mono WAV file to [path] at [sampleRate].
  static void writeWav(String path, List<double> samples, int sampleRate) {
    final bytes = writeWavBytes(samples, sampleRate: sampleRate);
    File(path).writeAsBytesSync(bytes);
  }
}

File: UI/Android/app/src/main/java/com/spproject/audiowatermark/WavUtils.kt
package com.spproject.audiowatermark

import android.content.Context
import java.io.File
import java.io.FileOutputStream
import java.io.InputStream

/**
 * WAV file I/O — reads and writes 44100 Hz / mono / 16-bit PCM.
 *
 * Required host-song format:
 *   44100 Hz, 16-bit PCM WAV (mono or stereo).
 */
object WavUtils {

    // ── Load ─────────────────────────────────────────────────────────────

    /**
     * Loads a WAV resource from res/raw into a normalized [-1.0, 1.0] DoubleArray.
     */
    fun loadWavAsDoubles(context: Context, resId: Int): DoubleArray {
        val bytes = context.resources.openRawResource(resId).use { it.readBytes() }
        return parseWavBytes(bytes)
    }

    /**
     * Loads a WAV from any [InputStream] into a normalized [-1.0, 1.0] DoubleArray.
     */
    fun loadWavFromStreamAsDoubles(inputStream: InputStream): DoubleArray {
        val bytes = inputStream.use { it.readBytes() }
        return parseWavBytes(bytes)
    }

    /**
     * Parses RIFF/WAVE PCM bytes into normalized [-1.0, 1.0] doubles.
     */
    private fun parseWavBytes(bytes: ByteArray): DoubleArray {
        require(bytes.size >= 12 &&
                String(bytes, 0, 4, Charsets.US_ASCII) == "RIFF" &&
                String(bytes, 8, 4, Charsets.US_ASCII) == "WAVE") {
            "Not a valid RIFF/WAVE file"
        }

        var pos = 12
        var numChannels = 1
        var sampleRate = 0
        var bitsPerSample = 16
        var dataOffset = -1
        var dataSize = -1

        while (pos + 8 <= bytes.size) {
            val chunkId = String(bytes, pos, 4, Charsets.US_ASCII)
            val chunkSize = readLE32(bytes, pos + 4)
            when (chunkId) {
                "fmt " -> {
                    val audioFormat = readLE16(bytes, pos + 8)   // 1 = PCM
                    require(audioFormat == 1) {
                        "Only uncompressed PCM WAV is supported (audioFormat=$audioFormat)."
                    }
                    numChannels = readLE16(bytes, pos + 10)
                    sampleRate  = readLE32(bytes, pos + 12)
                    bitsPerSample = readLE16(bytes, pos + 22)
                }
                "data" -> {
                    dataOffset = pos + 8
                    dataSize   = chunkSize
                }
            }
            pos += 8 + chunkSize + (chunkSize and 1)
            if (dataOffset != -1) break
        }

        require(dataOffset != -1)  { "No 'data' chunk found in WAV." }
        require(bitsPerSample == 16) {
            "Expected 16-bit PCM, got ${bitsPerSample}-bit."
        }
        require(sampleRate == Config.SAMPLE_RATE) {
            "Audio file is $sampleRate Hz, but app expects ${Config.SAMPLE_RATE} Hz."
        }

        val totalFrames = dataSize / (numChannels * 2)   // 2 bytes per 16-bit sample
        val doubles = DoubleArray(totalFrames)
        var bi = dataOffset
        for (i in 0 until totalFrames) {
            if (numChannels == 1) {
                doubles[i] = readLE16Signed(bytes, bi) / 32768.0
                bi += 2
            } else {
                // Downmix stereo → mono by averaging L and R channels
                val left  = readLE16Signed(bytes, bi)
                val right = readLE16Signed(bytes, bi + 2)
                doubles[i] = (left + right) / 2.0 / 32768.0
                bi += numChannels * 2
            }
        }
        return doubles
    }

    // ── Write ─────────────────────────────────────────────────────────────

    /**
     * Writes normalized [-1.0, 1.0] [samples] to a 44100 Hz / mono / 16-bit
     * PCM WAV file at [outFile].
     */
    fun writeWav(samples: DoubleArray, outFile: File) {
        val numSamples = samples.size
        val byteRate   = Config.SAMPLE_RATE * 2         // 1 ch × 2 bytes/sample
        val dataSize   = numSamples * 2                 // bytes in data chunk
        val fileSize   = 36 + dataSize                  // total RIFF size - 8

        FileOutputStream(outFile).use { fos ->
            // RIFF header
            fos.write("RIFF".toByteArray(Charsets.US_ASCII))
            fos.writeLE32(fileSize)
            fos.write("WAVE".toByteArray(Charsets.US_ASCII))

            // fmt  chunk (16 bytes)
            fos.write("fmt ".toByteArray(Charsets.US_ASCII))
            fos.writeLE32(16)                            // chunk size
            fos.writeLE16(1)                             // PCM
            fos.writeLE16(1)                             // mono
            fos.writeLE32(Config.SAMPLE_RATE)
            fos.writeLE32(byteRate)
            fos.writeLE16(2)                             // block align
            fos.writeLE16(16)                            // bits per sample

            // data chunk
            fos.write("data".toByteArray(Charsets.US_ASCII))
            fos.writeLE32(dataSize)
            val buf = ByteArray(2)
            for (s in samples) {
                val v = (s * 32767.0).toInt().coerceIn(-32768, 32767)
                buf[0] = (v and 0xFF).toByte()
                buf[1] = ((v shr 8) and 0xFF).toByte()
                fos.write(buf)
            }
        }
    }

    // ── LE read helpers ──────────────────────────────────────────────────

    private fun readLE16(b: ByteArray, offset: Int): Int =
        (b[offset].toInt() and 0xFF) or ((b[offset + 1].toInt() and 0xFF) shl 8)

    private fun readLE16Signed(b: ByteArray, offset: Int): Int {
        val lo = b[offset].toInt() and 0xFF
        val hi = b[offset + 1].toInt()
        return (hi shl 8) or lo
    }

    private fun readLE32(b: ByteArray, offset: Int): Int =
        (b[offset].toInt() and 0xFF) or
        ((b[offset + 1].toInt() and 0xFF) shl 8) or
        ((b[offset + 2].toInt() and 0xFF) shl 16) or
        ((b[offset + 3].toInt() and 0xFF) shl 24)

    // ── LE write helpers ─────────────────────────────────────────────────

    private fun FileOutputStream.writeLE16(v: Int) {
        write(v and 0xFF)
        write((v shr 8) and 0xFF)
    }

    private fun FileOutputStream.writeLE32(v: Int) {
        write(v and 0xFF)
        write((v shr 8) and 0xFF)
        write((v shr 16) and 0xFF)
        write((v shr 24) and 0xFF)
    }
}
